python/zfp.pyx

```diff
@@ -0,0 +1,423 @@
+        raise RuntimeError("error during zfp decompression")
+    return decomp_arr
+
+cdef _decompress_with_user_array(
+    zfp_field* field,
+    zfp_stream* stream,
+    void* out,
+):
+    with nogil:
+        zfp_field_set_pointer(field, out)
+        ret = zfp_decompress(stream, field)
+    if ret == 0:
+        raise RuntimeError("error during zfp decompression")
+
+cdef _set_compression_mode(
+    zfp_stream *stream,
+    zfp_type ztype,
+    int ndim,
+    double tolerance = -1,
+    double rate = -1,
+    int precision = -1,
+):
+    if tolerance >= 0:
+        zfp_stream_set_accuracy(stream, tolerance)
+    elif rate >= 0:
+        zfp_stream_set_rate(stream, rate, ztype, ndim, 0)
+    elif precision >= 0:
+        zfp_stream_set_precision(stream, precision)
+    else:
+        zfp_stream_set_reversible(stream)
+
+cdef _validate_4d_list(in_list, list_name):
+    # Validate that the input list is either a valid list for strides or shape
+    # Specifically, check it is a list and the length is > 0 and <= 4
+    # Throws a TypeError or ValueError if invalid
+    try:
+        if len(in_list) > 4:
+            raise ValueError(
+                "User-provided {} has too many dimensions "
+                "(up to 4 supported)"
+            )
+        elif len(in_list) <= 0:
+            raise ValueError(
+                    "User-provided {} needs at least one dimension"
+            )
+    except TypeError:
+        raise TypeError(
+            "User-provided {} is not an iterable"
+        )
+
+cdef _validate_userinput_matches_header(
+    zfp_field* field,
+    zfp_stream* stream,
+    out,
+    zfp_type ztype,
+    shape,
+    strides,
+    double tolerance,
+    double rate,
+    int precision
+):
+    # check that the header and user type matches
+    if ztype != type_none and ztype is not field[0]._type:
+        raise ValueError(
+            "User-provided zfp_type does not match zfp_type in header"
+        )
+
+    # check that the header and user shapes match
+    header_shape = (field[0].nx, field[0].ny, field[0].nz, field[0].nw)
+    header_shape = [x for x in header_shape if x > 0]
+    if shape is not None:
+        if not all([x == y for x, y in zip_longest(shape, header_shape)]):
+           raise ValueError(
+               "User-provided shape does not match shape in header"
+           )
+
+    # check that the shape and strides have the same number of dimensions
+    if shape is not None and sum(strides) > 0:
+        stride_dims = sum([1 for x in strides if x > 0])
+        shape_dims = sum([1 for x in shape if x > 0])
+        if len(strides) != len(shape):
+            raise ValueError(
+                "Mis-match in shape and stride lengths"
+            )
+
+    # check that setting the compression parameters based on user input
+    # does not change the stream mode (i.e., the compression parameters
+    # provided by the user and in the header match)
+    if tolerance >= 0 or rate >= 0 or precision >= 0:
```

Comment:
I think we also need to handle the case when none of these are true, i.e., when reversible mode is requested.

Reply:
This is a good point. We may have to rethink the API for `_decompress`.  How to differentiate between the user not specifying anything and the user intended to specify `reversible`?  Currently, there is ambiguity.

In the case where the user doesn't want to specify anything, we should check against the header.  In the case where the user does want to specify reversible, we should check.  Maybe we just add an additional default argument `reversible = False` and if the other three are `-1` and `reversible == False`, then don't check, else check the parameter that is set.

Reply:
From a use case standpoint, we'd ideally want the user to specify array metadata and compression parameters only when a header is not present.  When the header is present, I don't think we even want to allow the user to specify parameters.  All we could do then is verify that those parameters match the header.  If somehow the user already knows this information, then it made little sense to write it in a header in the first place.

Would it make sense to add a Boolean, read_header, that is the counterpart to the write_header Boolean specified in the compressor?  If it is true, then attempt to read the header and fail if not present.  If false, then assume there's no header and use the user-supplied compression parameters

Alternatively, instead of a Boolean argument, this could also be separated into two different functions (one that requires a header and one that forbids a header).  This would help avoid incompatible arguments.

I prefer this over adding another potentially incompatible parameter to specify that the user expects reversible compression.

Reply:
> Alternatively, instead of a Boolean argument, this could also be separated into two different functions (one that requires a header and one that forbids a header). This would help avoid incompatible arguments.

> I prefer this over adding another potentially incompatible parameter to specify that the user expects reversible compression.

Sure.  We already have a function that requires a header (`decompress_numpy`).  We could then change `_decompress` to forbid a header.  This will greatly simplify the `_decompress` function since it will no longer have two "personalities".  Do this sound good to you @lindstro?

If we go this route, should `_decompress` still attempt to read a header and throw an exception if it detects one?  Or should it just trust that the user provided a stream with no header?  I would normally lean towards extra checking, but I'm a little concerned about a *false positive* header detection (i.e., by chance the first bits of the headerless stream decode to `zfp`).  Thoughts?

Reply:
Great point!  Yes, let's simplify `_decompress` so that it does not read a header.  Since this function is not (yet) officially part of the API and is intended for experts only, I think it's fine to require the user to know what they're doing.

One downside of this design is that there is no way to first compress via `compress_numpy`, which outputs a header, and later decompress into a user-allocated buffer via `_decompress`, which now expects no header.  This suggests the need for a corresponding `_compress` function that does not write a header.

Reply:
> Great point! Yes, let's simplify _decompress so that it does not read a header. Since this function is not (yet) officially part of the API and is intended for experts only, I think it's fine to require the user to know what they're doing.

Sounds good. I'll make this change now.

> One downside of this design is that there is no way to first compress via compress_numpy, which outputs a header, and later decompress into a user-allocated buffer via _decompress, which now expects no header. This suggests the need for a corresponding _compress function that does not write a header.

The `compress_numpy` function has an optional `write_header` boolean argument to enable this use-case (as well as repeated compression to the same stream). The checksum tests make use of the `write_header` flag: https://github.com/LLNL/zfp/pull/29/files#diff-9fbf217f1e38984b823dcd123ec51ec8R64

Reply:
Of course.  Not sure what I was thinking.  Thanks for taking care of this issue.

Reply:
No worries!  While I'm refactoring the functions, do you mind if I remove `strides` as an argument to `_decompress`?  `strides` for decompression is currently broken in ways that I don't fully understand.  I think it may be better to withhold that functionality for now, and add it in a later release after it is tested and fully working. Thoughts?

Reply:
I can live with that.  Does this mean that strides will be inferred from the numpy array, when that's the target, or will otherwise be set to default values (sx = 1, sy = nx, sz = nx * ny, sw = nx * ny * nz)?

PS. One issue with strides that I pointed out earlier is that zfp measures strides in scalars while ndarray measures strides in bytes.

Reply:
> Does this mean that strides will be inferred from the numpy array, when that's the target, or will otherwise be set to default values (sx = 1, sy = nx, sz = nx * ny, sw = nx * ny * nz)?

I was thinking of leaving them unspecified (i.e., not modifying the stride values in the `field` so that they all remain 0 from initialization) and letting zfp calculate them.  Does that work?

PS - since the header is now assume to not exist, `shape` and `ztype` are no longer optional arguments to `_decompress`, they are now required arguments.

Reply:
> > Does this mean that strides will be inferred from the numpy array, when that's the target, or will otherwise be set to default values (sx = 1, sy = nx, sz = nx * ny, sw = nx * ny * nz)?
> 
> I was thinking of leaving them unspecified (i.e., not modifying the stride values in the `field` so that they all remain 0 from initialization) and letting zfp calculate them. Does that work?

I'm not opposed to that, but what happens if the user passes an array to decompress into?  Its strides will be ignored and not updated, right?  If those strides do not match what zfp assumes, then things will go wrong.  

> PS - since the header is now assume to not exist, `shape` and `ztype` are no longer optional arguments to `_decompress`, they are now required arguments.

Makes sense.  What happens if they do not match `out` when the user passes an array?

Reply:
> I'm not opposed to that, but what happens if the user passes an array to decompress into? Its strides will be ignored and not updated, right? If those strides do not match what zfp assumes, then things will go wrong.

Correct.  The user will have to do some massaging of the data/strides outside of ZFP to get things working if they don't want the default ZFP strides.

> Makes sense. What happens if they do not match out when the user passes an array?

Per our offline discussion, I'll refactor so that we continue to check that the out shape and type match if it is a numpy array, but also document that the user can pass in `array.data` as an "escape hatch" that avoids the check if they are confident they want to avoid it.

Reply:
This thread should be resolved as of the latest commit.